In [2]:
import graphlab
song_data = graphlab.SFrame('Downloads/song_data.gl/')

In [3]:
song_data.head()

Columns:
	user_id	str
	song_id	str
	listen_count	int
	title	str
	artist	str
	song	str

Rows: 10

Data:
+-------------------------------+--------------------+--------------+
|            user_id            |      song_id       | listen_count |
+-------------------------------+--------------------+--------------+
| b80344d063b5ccb3212f76538f... | SOAKIMP12A8C130995 |      1       |
| b80344d063b5ccb3212f76538f... | SOBBMDR12A8C13253B |      2       |
| b80344d063b5ccb3212f76538f... | SOBXHDL12A81C204C0 |      1       |
| b80344d063b5ccb3212f76538f... | SOBYHAJ12A6701BF1D |      1       |
| b80344d063b5ccb3212f76538f... | SODACBL12A8C13C273 |      1       |
| b80344d063b5ccb3212f76538f... | SODDNQT12A6D4F5F7E |      5       |
| b80344d063b5ccb3212f76538f... | SODXRTY12AB0180F3B |      1       |
| b80344d063b5ccb3212f76538f... | SOFGUAY12AB017B0A8 |      1       |
| b80344d063b5ccb3212f76538f... | SOFRQTD12A81C233C0 |      1       |
| b80344d063b5ccb3212f76538f... | SOHQWYZ12A6D4FA701 |      1       |
+-------------------------------+--------------------+--------------+
+-------------------------------+-------------------------------+
|             title             |             artist            |
+-------------------------------+-------------------------------+
|            The Cove           |          Jack Johnson         |
|        Entre Dos Aguas        |         Paco De Lucia         |
|            Stronger           |           Kanye West          |
|         Constellations        |          Jack Johnson         |
|          Learn To Fly         |          Foo Fighters         |
|  Apuesta Por El Rock 'N' Roll |      Héroes del Silencio      |
|         Paper Gangsta         |           Lady GaGa           |
|         Stacked Actors        |          Foo Fighters         |
|         Sehr kosmisch         |            Harmonia           |
| Heaven's gonna burn your eyes | Thievery Corporation feat.... |
+-------------------------------+-------------------------------+
+-------------------------------+
|              song             |
+-------------------------------+
|    The Cove - Jack Johnson    |
| Entre Dos Aguas - Paco De ... |
|     Stronger - Kanye West     |
| Constellations - Jack Johnson |
|  Learn To Fly - Foo Fighters  |
| Apuesta Por El Rock 'N' Ro... |
|   Paper Gangsta - Lady GaGa   |
| Stacked Actors - Foo Fighters |
|    Sehr kosmisch - Harmonia   |
| Heaven's gonna burn your e... |
+-------------------------------+
[10 rows x 6 columns]

In [4]:
graphlab.canvas.set_target('ipynb')

In [14]:
selected_users = song_data[song_data['artist'] =='Lady GaGa']


In [15]:
len(selected_users.unique())

4129

#Kanye west 3775
foo 3429
TS 6277
LG 4129
Kings Of Leon
William Tabbert
Sehr kosmisch - Harmonia



In [18]:
song_data.groupby(key_columns='artist', operations={'total_count': graphlab.aggregate.SUM('listen_count')}).sort('total_count',ascending=True)

Columns:
	artist	str
	total_count	int

Rows: 3375

Data:
+-------------------------------+-------------+
|             artist            | total_count |
+-------------------------------+-------------+
|        William Tabbert        |      14     |
|         Reel Feelings         |      24     |
| Beyoncé feat. Bun B and Sl... |      26     |
|         Boggle Karaoke        |      30     |
|             Diplo             |      30     |
|         harvey summers        |      31     |
|             Nâdiya            |      36     |
|          Jody Bernal          |      38     |
|        Aneta Langerova        |      38     |
| Kanye West / Talib Kweli /... |      38     |
+-------------------------------+-------------+
[3375 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [19]:
train_data,test_data = song_data.random_split(.8,seed=0)

In [20]:
personalized_model = graphlab.item_similarity_recommender.create(train_data,
                                                                user_id='user_id',
                                                                item_id='song')

Recsys training: model = item_similarity

Warning: Ignoring columns song_id, listen_count, title, artist;

To use one of these as a target column, set target = <column_name>

and use a method that allows the use of a target.

Preparing data set.

Data has 893580 observations with 66085 users and 9952 items.

Data prepared in: 0.670492s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.887ms                        | 6          |

| 20.165ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 196.482ms                           | 0                | 0               |

| 579.307ms                           | 100              | 9952            |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 1.62978s

In [21]:
subset_test_users = test_data['user_id'].unique()[0:10000]

In [22]:
personalized_model.recommend(subset_test_users,k=1)

recommendations finished on 1000/10000 queries. users per second: 22511.9

recommendations finished on 2000/10000 queries. users per second: 30019.7

recommendations finished on 3000/10000 queries. users per second: 29991.6

recommendations finished on 4000/10000 queries. users per second: 32784.5

recommendations finished on 5000/10000 queries. users per second: 33127.3

recommendations finished on 6000/10000 queries. users per second: 32895.3

recommendations finished on 7000/10000 queries. users per second: 33081.4

recommendations finished on 8000/10000 queries. users per second: 32575.9

recommendations finished on 9000/10000 queries. users per second: 33450.9

recommendations finished on 10000/10000 queries. users per second: 33271.3

Columns:
	user_id	str
	song	str
	score	float
	rank	int

Rows: 10000

Data:
+-------------------------------+-------------------------------+
|            user_id            |              song             |
+-------------------------------+-------------------------------+
| c067c22072a17d33310d7223d7... | Grind With Me (Explicit Ve... |
| 696787172dd3f5169dc94deef9... | Senza Una Donna (Without A... |
| 532e98155cbfd1e1a474a28ed9... | Jive Talkin' (Album Versio... |
| 18325842a941bc58449ee71d65... | Goodnight And Goodbye - Jo... |
| 507433946f534f5d25ad1be302... | Find The Cost Of Freedom -... |
| 18fafad477f9d72ff86f7d0bd8... | Rabbit Heart (Raise It Up)... |
| fe85b96ba1983219b296f6b486... |     Secrets - OneRepublic     |
| 225ea420b4bede50919d1bfe24... |       Clocks - Coldplay       |
| 95dc7e2b188b1148b2d25f4e6b... | Bust a Move - Infected Mus... |
| 4a3a1ae2748f12f7ab921a47d6... | Isis (Spam Remix) - Alaska... |
+-------------------------------+-------------------------------+
+-----------------+------+
|      score      | rank |
+-----------------+------+
| 0.0459424376488 |  1   |
|  0.017026577677 |  1   |
| 0.0118288653237 |  1   |
| 0.0159257985651 |  1   |
| 0.0165806589303 |  1   |
| 0.0799399726093 |  1   |
| 0.0788827141126 |  1   |
| 0.0271030251796 |  1   |
| 0.0534738540649 |  1   |
|  0.04180302118  |  1   |
+-----------------+------+
[10000 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [25]:
song_data.groupby(key_columns='song', operations={'count': graphlab.aggregate.COUNT()}).sort('count',ascending=False)

Columns:
	song	str
	count	int

Rows: 9952

Data:
+--------------------------------+-------+
|              song              | count |
+--------------------------------+-------+
|    Sehr kosmisch - Harmonia    |  5970 |
|          Undo - Björk          |  5281 |
| You're The One - Dwight Yoakam |  4806 |
| Dog Days Are Over (Radio E...  |  4536 |
|    Revelry - Kings Of Leon     |  4339 |
| Horn Concerto No. 4 in E f...  |  3949 |
|     Secrets - OneRepublic      |  3916 |
|       Tive Sim - Cartola       |  3185 |
| Fireflies - Charttraxx Karaoke |  3171 |
|    Hey_ Soul Sister - Train    |  3132 |
+--------------------------------+-------+
[9952 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [26]:
train_data,test_data = song_data.random_split(.8,seed=0)

In [27]:
personalized_model = graphlab.item_similarity_recommender.create(train_data,
                                                                user_id='user_id',
                                                                item_id='song')

Recsys training: model = item_similarity

Warning: Ignoring columns song_id, listen_count, title, artist;

To use one of these as a target column, set target = <column_name>

and use a method that allows the use of a target.

Preparing data set.

Data has 893580 observations with 66085 users and 9952 items.

Data prepared in: 0.684128s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.701ms                        | 1.5        |

| 20.04ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 191.972ms                           | 0                | 0               |

| 730.838ms                           | 100              | 9952            |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 0.804092s

In [28]:
subset_test_users = test_data['user_id'].unique()[0:10000]

In [29]:
personalized_model.recommend(subset_test_users,k=1)

recommendations finished on 1000/10000 queries. users per second: 24692

recommendations finished on 2000/10000 queries. users per second: 29905.2

recommendations finished on 3000/10000 queries. users per second: 32670.1

recommendations finished on 4000/10000 queries. users per second: 30780.6

recommendations finished on 5000/10000 queries. users per second: 33430.7

recommendations finished on 6000/10000 queries. users per second: 35517.9

recommendations finished on 7000/10000 queries. users per second: 35650.4

recommendations finished on 8000/10000 queries. users per second: 35020.3

recommendations finished on 9000/10000 queries. users per second: 35251.8

recommendations finished on 10000/10000 queries. users per second: 32536.8

Columns:
	user_id	str
	song	str
	score	float
	rank	int

Rows: 10000

Data:
+-------------------------------+-------------------------------+
|            user_id            |              song             |
+-------------------------------+-------------------------------+
| c067c22072a17d33310d7223d7... | Grind With Me (Explicit Ve... |
| 696787172dd3f5169dc94deef9... | Senza Una Donna (Without A... |
| 532e98155cbfd1e1a474a28ed9... | Jive Talkin' (Album Versio... |
| 18325842a941bc58449ee71d65... | Goodnight And Goodbye - Jo... |
| 507433946f534f5d25ad1be302... | Find The Cost Of Freedom -... |
| 18fafad477f9d72ff86f7d0bd8... | Rabbit Heart (Raise It Up)... |
| fe85b96ba1983219b296f6b486... |     Secrets - OneRepublic     |
| 225ea420b4bede50919d1bfe24... |       Clocks - Coldplay       |
| 95dc7e2b188b1148b2d25f4e6b... | Bust a Move - Infected Mus... |
| 4a3a1ae2748f12f7ab921a47d6... | Isis (Spam Remix) - Alaska... |
+-------------------------------+-------------------------------+
+-----------------+------+
|      score      | rank |
+-----------------+------+
| 0.0459424376488 |  1   |
|  0.017026577677 |  1   |
| 0.0118288653237 |  1   |
| 0.0159257985651 |  1   |
| 0.0165806589303 |  1   |
| 0.0799399726093 |  1   |
| 0.0788827141126 |  1   |
| 0.0271030251796 |  1   |
| 0.0534738540649 |  1   |
|  0.04180302118  |  1   |
+-----------------+------+
[10000 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [31]:
song_data.groupby(key_columns='song', operations={'count': graphlab.aggregate.COUNT()}).sort('count',ascending=False)

Columns:
	song	str
	count	int

Rows: 9952

Data:
+--------------------------------+-------+
|              song              | count |
+--------------------------------+-------+
|    Sehr kosmisch - Harmonia    |  5970 |
|          Undo - Björk          |  5281 |
| You're The One - Dwight Yoakam |  4806 |
| Dog Days Are Over (Radio E...  |  4536 |
|    Revelry - Kings Of Leon     |  4339 |
| Horn Concerto No. 4 in E f...  |  3949 |
|     Secrets - OneRepublic      |  3916 |
|       Tive Sim - Cartola       |  3185 |
| Fireflies - Charttraxx Karaoke |  3171 |
|    Hey_ Soul Sister - Train    |  3132 |
+--------------------------------+-------+
[9952 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.